Default to padding blinded paths

TheBlueMatt opened the pull request:
After much discussion in #3246 we mostly decided to allow
    downstream developers to override whatever decisions the
    `DefaultMessageRouter` makes regarding blinded path selection by
    providing easy overrides for the selected `OnionMessageRouter`. We
    did not, however, actually select good defaults for
    `DefaultMessageRouter`.

    Here we add those defaults, taking advantage of the
    `MessageContext` we're given to detect why we're building a blinded
    path and selecting blinding and compaction parameters based on it.

    Specifically, if the blinded path is not being built for an offers
    context, we always use a non-compact blinded path and always pad it
    to four hops (including the recipient).

    However, if the blinded path is being built for an `Offers` context
    which implies it might need to fit in a QR code (or, worse, a
    payment onion), we reduce our padding and try to build a compact
    blinded path if possible.

    We retain the `NodeIdMessageRouter` to disable compact blinded path
    creation but use the same path-padding heuristic as for
    `DefaultMessageRouter`.

valentinewallace commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -800,7 +829,7 @@ where
 			&self.entropy_source,
 			secp_ctx,
-			false,
+			true,
```
It's confusing that the docs on the struct say the `NodeIdMessageRouter` always creates non-compact paths, but we set `compact_paths` to `true` here. Can you add a comment?

valentinewallace commented on lightning/src/blinded_path/message.rs:
```diff
@@ -737,18 +759,42 @@ pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
+
+	let round_off = if compact_padding {
+		// We can only pad by a minimum of two bytes. Thus, if there are any intermediate hops that
+		// need to be padded by exactly one byte, we have to instead pad everything by two.
```
Not sure there's a clear solution, but "padding" in this PR can refer to either whole-path padding (i.e. dummy hops) or per-hop padding (i.e. making (at least some of) the hops equal size), and it's a bit confusing. Would prefer using the dummy hop terminology for whole-path padding.

valentinewallace commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -755,7 +780,8 @@ where
-/// message, and thus an `Err` is returned.
+/// message, and thus an `Err` is returned. The impact of this may be somewhat muted when
+/// additional padding is added to the blinded path, but this protection is not complete.
 pub struct NodeIdMessageRouter<G: Deref<Target = NetworkGraph<L>>, L: Deref, ES: Deref>
```
Not related to the PR per se, I'm wondering if we can get rid of this struct? I don't see it used anywhere and the docs don't indicate when it should be used

TheBlueMatt commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -163,6 +163,17 @@ fn check_compact_path_introduction_node<'a, 'b, 'c>(
+	expected_path_length: usize,
+) -> bool {
+	let introduction_node_id = resolve_introduction_node(lookup_node, path);
+	introduction_node_id == expected_introduction_node
```
We quite often use non-compact paths now, as we generally try to use them any time where we're not space-constrained, so such a check fails a handful of test.

TheBlueMatt commented on lightning/src/blinded_path/message.rs:
```diff
@@ -737,18 +759,42 @@ pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
+		intermediate_tlvs.clone().any(|tlvs| tlvs.serialized_length() == max_intermediate_len - 1);
+
+	let round_off = if compact_padding {
+		// We can only pad by a minimum of two bytes. Thus, if there are any intermediate hops that
```
Its a TLV - type and length are two bytes.

valentinewallace commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -703,32 +694,70 @@ fn test_blinded_path_padding_for_full_length_path() {
+
+#[test]
+fn test_compact_blinded_path_compact_padding() {
+	// Check that for a blinded path with compact padding, no extra padding is applied.
```
nit: this test summary is the same as the above test's summary

TheBlueMatt commented:
Squashed and updated a few docs/comments:

```diff
$ git diff-tree -U2 649ba3478 8efb3e171
diff --git a/lightning/src/onion_message/functional_tests.rs b/lightning/src/onion_message/functional_tests.rs
index ffdb2f6403..75e2aaf3c5 100644
--- a/lightning/src/onion_message/functional_tests.rs
+++ b/lightning/src/onion_message/functional_tests.rs
@@ -696,5 +696,6 @@ fn test_blinded_path_padding_for_full_length_path() {
 #[test]
 fn test_blinded_path_compact_padding() {
-	// Check that for a blinded path with compact padding, no extra padding is applied.
+	// Check that for a blinded path with non-SCID intermediate hops with compact padding, no extra
+	// padding is applied.
 	let nodes = create_nodes(4);

@@ -729,5 +730,6 @@ fn test_blinded_path_compact_padding() {
 #[test]
 fn test_compact_blinded_path_compact_padding() {
-	// Check that for a blinded path with compact padding, no extra padding is applied.
+	// Check that for a blinded path with SCID intermediate hops with compact padding, no extra
+	// padding is applied.
 	let nodes = create_nodes(4);

diff --git a/lightning/src/onion_message/messenger.rs b/lightning/src/onion_message/messenger.rs
index d6d4314196..03fc8dd4f7 100644
--- a/lightning/src/onion_message/messenger.rs
+++ b/lightning/src/onion_message/messenger.rs
@@ -773,4 +773,7 @@ where
 /// paths.
 ///
+/// This may be useful in cases where you want a long-lived blinded path and anticipate channel(s)
+/// may close, but connections to specific peers will remain stable.
+///
 /// This message router can only route to a directly connected [`Destination`].
 ///
```

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -54,21 +54,38 @@ impl Readable for BlindedMessagePath {
 	/// Create a one-hop blinded path for a message.
+	///
+	/// `compact_padding` selects between space-inefficient padding that better hides contents and
+	/// a space-constrained padding that does very little to hide the contents, especially for the
```
s/that/which

TheBlueMatt commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -163,6 +163,20 @@ fn check_compact_path_introduction_node<'a, 'b, 'c>(
 		&& matches!(path.introduction_node(), IntroductionNode::DirectedShortChannelId(..))
 }
 
+fn check_dummy_hopd_path_length<'a, 'b, 'c>(
```
Makes it sound like the dummy is hopping over the path, but sure :joy:

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -737,18 +759,43 @@ pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
+
+	let round_off = if compact_padding {
+		// We can only pad by a minimum of two bytes (we can only go from no-TLV to a type + length
+		// byte). Thus, if there are any intermediate hops that need to be padded by exactly one
```
> The point is we can never pad by a single byte, so if we need to pad by a single byte, we have to instead just pad everything by 1/2.

I guess that's what I don't understand. I get that a TLV of length zero would be only two bytes. But why is that fact important when padding? Won't all `ForwardTlvs` have the same length even without padding while `DummyTlvs` serialize to an empty TLV (i.e., two bytes)? So in practice, aren't we simply padding to the serialized `ForwardTlvs` size (i.e., two bytes for type + length plus an 8-byte SCID)?

jkczyz reviewed (commented):
Silencing the review bot since my review was a single comment on an existing discussion.
